nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,323 @@
+							Description: "The username for login",
+							Required:    true,
+						},
+					},
+				},
+				Required: true,
+			},
+		},
+	}
+}
+
+type NsxClusterNode struct {
+	ID        *string
+	IPAddress *string
+	UserName  *string
+	Password  *string
+}
+
+func getClusterNodesFromSchema(d *schema.ResourceData) []NsxClusterNode {
+	nodes := d.Get("nodes").([]interface{})
+	var clusterNodes []NsxClusterNode
+	for _, node := range nodes {
+		data := node.(map[string]interface{})
+		id := data["id"].(string)
+		ipAddress := data["ip_address"].(string)
+		userName := data["username"].(string)
+		password := data["password"].(string)
+		nodeObj := NsxClusterNode{
+			ID:        &id,
+			IPAddress: &ipAddress,
+			UserName:  &userName,
+			Password:  &password,
+		}
+		clusterNodes = append(clusterNodes, nodeObj)
+	}
+	return clusterNodes
+}
+
+func setNodesInSchema(d *schema.ResourceData, nodes []NsxClusterNode) error {
+	nodeIPToCredentialMap := getIPtoCredentialMap(d)
+	var nodeList []map[string]interface{}
+	for _, node := range nodes {
+		elem := make(map[string]interface{})
+		elem["id"] = node.ID
+		elem["ip_address"] = node.IPAddress
+		elem["username"] = nodeIPToCredentialMap[*node.IPAddress][0]
+		elem["password"] = nodeIPToCredentialMap[*node.IPAddress][1]
+		nodeList = append(nodeList, elem)
+	}
+	return d.Set("nodes", nodeList)
+}
+
+func getIPtoCredentialMap(d *schema.ResourceData) map[string][]string {
+	// returns a map, whose key is IP address of a node,
+	// value is a slice of [node_login_username, node_login_password]
+	nodes := d.Get("nodes").([]interface{})
+	var idToCredentialMap = map[string][]string{}
+	for _, node := range nodes {
+		data := node.(map[string]interface{})
+		ipAddress := data["ip_address"].(string)
+		userName := data["username"].(string)
+		password := data["password"].(string)
+		cred := []string{userName, password}
+		idToCredentialMap[ipAddress] = cred
+	}
+	return idToCredentialMap
+}
+
+func resourceNsxtManagerClusterCreate(d *schema.ResourceData, m interface{}) error {
+	// Call Joincluster function on nodes that are not in the cluster
+	nodes := getClusterNodesFromSchema(d)
+	if len(nodes) == 0 {
+		return fmt.Errorf("Please provide at least one node ip")
+	} else if len(nodes) == 1 {
+		log.Printf("[INFO] Only one node ip provided, skip joining nodes process")
+		return resourceNsxtManagerClusterRead(d, m)
+	}
+	hostIP, _ := getHostIPFromNsxtClient(m)
+	clusterID, certSha256Thumbprint, err := getClusterInfoFromHostNode(d, m)
+	if err != nil {
+		return err
+	}
+
+	for _, guestNode := range nodes {
+		err := joinNodeToCluster(clusterID, certSha256Thumbprint, guestNode, hostIP, d, m)
+		if err != nil {
+			return err
+		}
+	}
+	d.SetId(clusterID)
+	d.Set("cluster_id", clusterID)
+	return resourceNsxtManagerClusterRead(d, m)
+}
+
+func getHostIPFromNsxtClient(client interface{}) (string, bool) {
+	c := client.(nsxtClients)
+	hostIP := c.Host
+	hostIP = strings.TrimPrefix(hostIP, "https://")
```

Comment:
This is not necessarily IP address. Can we resolve the host to IP programmatically if fqdn is specified?